File: src/main/java/com/bajdcc/util/lexer/automata/dfa/DFAStatusData.java
package com.bajdcc.util.lexer.automata.dfa;

import com.bajdcc.util.lexer.automata.nfa.NFAStatus;
import com.bajdcc.util.lexer.automata.nfa.NFAStatusData;

import java.util.ArrayList;
import java.util.Arrays;

/**
 * DFA状态数据
 *
 * @author bajdcc
 */
public class DFAStatusData extends NFAStatusData {
	/**
	 * NFA状态集合
	 */
	public ArrayList<NFAStatus> nfaStatus = new ArrayList<>();

	/**
	 * 获得状态编号描述（逗号分隔）
	 *
	 * @param dfaStatusList 状态表
	 * @return 状态编号描述
	 */
	public String getStatusString(ArrayList<NFAStatus> dfaStatusList) {
		if (nfaStatus.isEmpty()) {
			return "";
		}
		int[] orders = new int[nfaStatus.size()];
		for (int i = 0; i < orders.length; i++) {
			orders[i] = dfaStatusList.indexOf(nfaStatus.get(i));
		}
		Arrays.sort(orders);
		StringBuilder sb = new StringBuilder();
		for (int order : orders) {
			sb.append(order).append(",");
		}
		return sb.toString();
	}
}

File: src/main/java/com/bajdcc/util/lexer/automata/dfa/DFAStatusData.kt
package com.bajdcc.util.lexer.automata.dfa

import com.bajdcc.util.lexer.automata.nfa.NFAStatus
import java.util.*

/**
 * DFA状态数据
 * [bFinal] 是否为终态
 * [nfaStatus] NFA状态集合
 * @author bajdcc
 */
data class DFAStatusData(var bFinal: Boolean = false,
                         var nfaStatus: MutableList<NFAStatus> = mutableListOf()) {
    /**
     * 获得状态编号描述（逗号分隔）
     *
     * @param dfaStatusList 状态表
     * @return 状态编号描述
     */
    fun getStatusString(dfaStatusList: List<NFAStatus>): String {
        if (nfaStatus.isEmpty()) {
            return ""
        }
        val orders = IntArray(nfaStatus.size)
        for (i in orders.indices) {
            orders[i] = dfaStatusList.indexOf(nfaStatus[i])
        }
        Arrays.sort(orders)
        val sb = StringBuilder()
        for (order in orders) {
            sb.append(order).append(",")
        }
        return sb.toString()
    }
}
